src/Vendr.PaymentProviders.Worldpay/WorldpayPaymentProvider.cs

```diff
@@ -1,38 +1,139 @@
+using System.Linq;
 using System.Web;
 using System.Web.Mvc;
 using Vendr.Core;
+using Vendr.Core.Logging;
 using Vendr.Core.Models;
+using Vendr.Core.Services;
 using Vendr.Core.Web.Api;
 using Vendr.Core.Web.PaymentProviders;
+using Vendr.PaymentProviders.Worldpay.Helpers;
 
-namespace Vendr.PaymentProviders
+namespace Vendr.PaymentProviders.Worldpay
 {
-    [PaymentProvider("worldpay", "Worldpay", "Worldpay payment provider", Icon = "icon-invoice")]
+    [PaymentProvider("worldpay", "Worldpay", "Worldpay payment provider", Icon = "icon-credit-card")]
     public class WorldpayPaymentProvider : PaymentProviderBase<WorldpaySettings>
     {
-        public WorldpayPaymentProvider(VendrContext vendr)
-            : base(vendr)
-        { }
+        private readonly ILogger _logger;
 
-        public override bool FinalizeAtContinueUrl => true;
+        public WorldpayPaymentProvider(VendrContext vendr, ILogger logger) : base(vendr)
+        {
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+        public override bool FinalizeAtContinueUrl => false;
 
         public override PaymentFormResult GenerateForm(OrderReadOnly order, string continueUrl, string cancelUrl, string callbackUrl, WorldpaySettings settings)
         {
-            return new PaymentFormResult()
+            try
+            {
+                if (settings.VerboseLogging)
+                {
+                    _logger.Info<WorldpayPaymentProvider>($"GenerateForm method called for cart {order.CartNumber}");
+                }
+
+                var url = settings.Mode.ToLower() == "live" ? settings.LiveUrl : settings.TestUrl;
+                var form = new PaymentForm(url, FormMethod.Post);
+
+                settings.InstallId.MustNotBeNull("settings.InstallId");
+                settings.TestModeNumber.MustNotBeNull("settings.TestModeNumber");
+                settings.AuthMode.MustNotBeNull("settings.AuthMode");
+                settings.OrderPropertyBillingFirstName.MustNotBeNull("settings.OrderPropertyBillingFirstName");
+                settings.OrderPropertyBillingLastName.MustNotBeNull("settings.OrderPropertyBillingLastName");
+
+                var firstname = order.Properties[settings.OrderPropertyBillingFirstName];
+                var surname = order.Properties[settings.OrderPropertyBillingLastName];
+                var address1 = order.Properties[settings.OrderPropertyBillingAddress1] ?? string.Empty;
+                var city = order.Properties[settings.OrderPropertyBillingCity] ?? string.Empty;
+                var postcode = order.Properties[settings.OrderPropertyBillingPostcode] ?? string.Empty;
+                var billingCountry = Vendr.Services.CountryService.GetCountry(order.PaymentInfo.CountryId.Value);
+                var billingCountryCode = billingCountry.Code.ToUpperInvariant();
+                var amount = order.TotalPrice.Value.WithTax.ToString("0.00", CultureInfo.InvariantCulture);
```

Comment:
Since Vendr 1.4.0 we now have a `TransactionAmount` property on the order which should be the amount used by payment providers. See this blog post for more https://vendr.net/blog/vendr-1-4-0-release-candidate/